### ISSUE TITLE
PHP cannot make concurrent calls

### ISSUE THREAD

**Issue by michaelbausor** [grpc/grpc#12175]
When making Unary method calls using PHP gRPC, it is not possible to start a call without also blocking on it, even though the UnaryCall object provides separate `start()` and `wait()` methods.

For example, the following Speech code:
```php
    $startTime = microtime(true);
    $call = $stub->Recognize($request, [], ['call_credentials_callback' => $tokenCallback]);
    $callTime = microtime(true);
    $result = $call->wait();
    $endTime = microtime(true);
    echo "Time to call: " . ($callTime - $startTime) . PHP_EOL;
    echo "Time to wait: " . ($endTime - $callTime) . PHP_EOL;
```
produces this output:
```
    Time to call: 1.3245630264282
    Time to wait: 0.0020620822906494
```
(see full code here: https://gist.github.com/michaelbausor/d3baa9db9e981949b0eded149a515e23)
The expected output would be that `Time to call` would be only the time required for gRPC to start the call, and that `Time to wait` would be the much longer time required for the Speech API to process the request.

Currently, this makes it impossible to start multiple calls in parallel. For example, this code:
```php
$requests = [*Array of N speech recognize requests*]
$calls = []
foreach ($requests as $request) {
  $calls[] = $stub->Recognize($request, [], ['call_credentials_callback' => $tokenCallback]);
}
foreach ($calls as $call) {
  $call->wait();
}
```
will always execute N requests in series, and thus take time equal to `N * average_request_length`. The expected behaviour would be that this would take time approximately equal to the time taken by the longest request.

This seems to be caused because currently the UnaryCall object sends the `OP_RECV_INITIAL_METADATA` code in the `start()` method. (See https://github.com/grpc/grpc/blob/master/src/php/lib/Grpc/UnaryCall.php#L45). This causes calls to the `start()` method to block.

Moving the `OP_RECV_INITIAL_METADATA` code from the `start()` method to the `wait()` method seems to resolve this issue.

cc @stanley-cheung @dwsupplee @jdpedrie @bshaffer

**Comment by stanley-cheung** [comment]
cc @murgatroid99

**Comment by murgatroid99** [comment]
One thing to note regarding those measurements is that a full second is an extreme amount of time for a single call, which probably means that this is the first call you're making with a channel. If that is the case, then the call time you're measuring includes the time it takes to establish a connection. I expect that you will get somewhat different results if you make a call before you start measuring.

Regardless of that, there would be some benefit to moving `OP_RECV_INITIAL_METADATA` to the `wait` method as you suggest. If there are only send operations in the `start` method, then it will usually not have to wait for a network round trip. However, the current unary call API guarantees that once the call has been started, the metadata is present.

One possible solution would be to have the `getMetadata` method itself call the `OP_RECV_INITIAL_METADATA` code if it has not already been read.

**Comment by michaelbausor** [comment]
@murgatroid99 Repeating the call does reduce the time a small amount (to about 0.8-0.9 seconds), but not significantly - see results below for looping 10 times.

In general, machine learning APIs such as Speech can make very long calls, with the duration being entirely dependent on the input. The synchronous recognition request for Speech can accept audio of up to around 1 minute in length, and according to the docs it is expected that "30 seconds of audio is processed in 15 seconds on average" ([see here](https://cloud.google.com/speech/docs/basics)).

More significantly though is the second example, of making multiple calls in parallel. Using Pubsub instead of Speech, with the current gRPC code it takes approximately 20 seconds to make 1000 calls to the `Publish` method. Making the modification of moving `OP_RECV_INITIAL_METADATA` reduces this to around 1.6 seconds.

Having `getMetadata` make the call makes a lot of sense to me. We could then also conditionally include the `OP_RECV_INITIAL_METADATA` flag in `wait()`, similar to what is done for `BidiStreamingCall::read()` (https://github.com/grpc/grpc/blob/master/src/php/lib/Grpc/BidiStreamingCall.php#L49)

Output of looping Speech test 10 times:
Time to call: 1.2568538188934
Time to wait: 0.0019960403442383
Time to call: 0.90999507904053
Time to wait: 0.00068497657775879
Time to call: 0.89781498908997
Time to wait: 0.00091099739074707
Time to call: 0.92992115020752
Time to wait: 0.00061488151550293
Time to call: 0.90489912033081
Time to wait: 0.0004119873046875
Time to call: 0.91139006614685
Time to wait: 0.00066804885864258
Time to call: 0.91404891014099
Time to wait: 0.00060796737670898
Time to call: 0.93215799331665
Time to wait: 0.00067496299743652
Time to call: 0.90055298805237
Time to wait: 0.00074911117553711
Time to call: 0.9248321056366
Time to wait: 0.00066709518432617

**Comment by michaelbausor** [comment]
I have created a PR that implements your suggestion for UnaryCall and ServerStreamingCall. Please take a look and let me know if this is what you had in mind: https://github.com/grpc/grpc/pull/12188

**Comment by vkedia** [comment]
WIth the proposed change, what happens if I call `start` but never call `wait`? Will the RPC still finish on the server?

**Comment by dwsupplee** [comment]
@vkedia In my testing of the proposed PR calls that simply execute `start` do in fact finish on the server - wait is only necessary to view the results of said call.

**Comment by stanley-cheung** [comment]
This is now included in `v1.4.7`
